Support non-ephemeral, self-hosted runners

subsequent builds fail with:

```
Run webfactory/ssh-agent@v0.2.00s
##[error]Node run failed with exit code 1
Run webfactory/ssh-agent@v0.2.0
Adding GitHub.com keys to undefined/.ssh/known_hosts
Starting ssh-agent
bind: Address already in use
unix_listener: cannot bind to path: /tmp/ssh-auth.sock
##[error]Command failed: ssh-agent -a /tmp/ssh-auth.sock
bind: Address already in use
unix_listener: cannot bind to path: /tmp/ssh-auth.sock

##[error]Node run failed with exit code 1
```

This is on a local github-runners, which are not ephemeral. Cleaning up the /tmp/ssh-auth.sock manually resolves this problem temporarily.

--- a/README.md
+++ b/README.md
@@ -130,7 +130,7 @@ As a note to my future self, in order to work on this repo:
 * Run `npm install` to fetch dependencies
 * _hack hack hack_
 * `node index.js`. Inputs are passed through `INPUT_` env vars with their names uppercased. Use `env "INPUT_SSH-PRIVATE-KEY=\`cat file\`" node index.js` for this action.
-* Run `./node_modules/.bin/ncc build index.js` to update `dist/index.js`, which is the file actually run
+* Run `npm run build` to update `dist/*`, which is the file actually run
 * Read https://help.github.com/en/articles/creating-a-javascript-action if unsure.
 * Maybe update the README example when publishing a new version.
 

--- a/README.md
+++ b/README.md
@@ -60,16 +60,7 @@ six different keys loaded into the `ssh-agent`, but the server aborts after five
 ## Exported variables
 The action exports `SSH_AUTH_SOCK` and `SSH_AGENT_PID` through the Github Actions core module.
 The `$SSH_AUTH_SOCK` is used by several applications like git or rsync to connect to the SSH authentication agent.
-The `$SSH_AGENT_PID` can be used to kill the SSH agent, e.g. when using on a self-hosted runner:
-```yaml
-jobs:
-    job_1:
-        steps:
-            # ...
-            -   name: Stop SSH agent
-                run: kill $SSH_AGENT_PID
-```
-The variables are only available inside the current job. 
+The `$SSH_AGENT_PID` contains the process id of the agent. This is used to kill the agent in post job action.
 
 ## Known issues and limitations
 
@@ -130,7 +121,7 @@ As a note to my future self, in order to work on this repo:
 * Run `npm install` to fetch dependencies
 * _hack hack hack_
 * `node index.js`. Inputs are passed through `INPUT_` env vars with their names uppercased. Use `env "INPUT_SSH-PRIVATE-KEY=\`cat file\`" node index.js` for this action.
-* Run `npm run build` to update `dist/*`, which is the file actually run
+* Run `npm run build` to update `dist/*`, which holds the files actually run
 * Read https://help.github.com/en/articles/creating-a-javascript-action if unsure.
 * Maybe update the README example when publishing a new version.
 

--- a/README.md
+++ b/README.md
@@ -57,6 +57,20 @@ The `ssh-agent` will load all of the keys and try each one in order when establi
 There's one **caveat**, though: SSH servers may abort the connection attempt after a number of mismatching keys have been presented. So if, for example, you have
 six different keys loaded into the `ssh-agent`, but the server aborts after five unknown keys, the last key (which might be the right one) will never even be tried.
 
+## Exported variables
+The action exports `SSH_AUTH_SOCK` and `SSH_AGENT_PID` through the Github Actions core module.
+The `$SSH_AUTH_SOCK` is used by several applications like git or rsync to connect to the SSH authentication agent.
+The `$SSH_AGENT_PID` can be used to kill the SSH agent, e.g. when using on a self-hosted runner:
+```yaml
+jobs:
+    job_1:
+        steps:
+            # ...
+            -   name: Stop SSH agent
+                run: kill $SSH_AGENT_PID
+```
+The variables are only available inside the current job. 
+
 ## Known issues and limitations
 
 ### Currently OS X and Linux only

--- a/action.yml
+++ b/action.yml
@@ -6,7 +6,6 @@ inputs:
         required: true
     ssh-auth-sock:
         description: 'Where to place the SSH Agent auth socket'
-        default: /tmp/ssh-auth.sock
 runs:
     using: 'node12'
     main: 'dist/index.js'

--- a/action.yml
+++ b/action.yml
@@ -9,6 +9,7 @@ inputs:
 runs:
     using: 'node12'
     main: 'dist/index.js'
+    post: 'dist/cleanup.js'
 branding:
     icon: loader
     color: 'yellow'

--- a/cleanup.js
+++ b/cleanup.js
@@ -1,0 +1,10 @@
+const core = require('@actions/core')
+const { execSync } = require('child_process')
+
+try {
+    // Kill the started SSH agent
+    console.log('Stopping SSH agent')
+    execSync('kill ${SSH_AGENT_PID}', { stdio: 'inherit' })
+} catch (error) {
+    core.setFailed(error.message)
+}

--- a/index.js
+++ b/index.js
@@ -22,8 +22,21 @@ try {
 
     console.log("Starting ssh-agent");
     const authSock = core.getInput('ssh-auth-sock');
-    child_process.execFileSync('ssh-agent', ['-a', authSock]);
-    core.exportVariable('SSH_AUTH_SOCK', authSock);
+    let sshAgentOutput = ''
+    if (authSock && authSock.length > 0) {
+        sshAgentOutput = child_process.execFileSync('ssh-agent', ['-a', authSock]);
+    } else {
+        sshAgentOutput = child_process.execFileSync('ssh-agent')
+    }
+
+    // Extract auth socket path and agent pid and set them as job variables
+    const lines = sshAgentOutput.toString().split("\n")
+    for (const lineNumber in lines) {
+        const matches = /^(SSH_AUTH_SOCK|SSH_AGENT_PID)=(.*); export \1/.exec(lines[lineNumber])
+        if (matches && matches.length > 0) {
+            core.exportVariable(matches[1], matches[2])
+        }
+    }
 
     console.log("Adding private key to agent");
     privateKey.split(/(?=-----BEGIN)/).forEach(function(key) {

--- a/package.json
+++ b/package.json
@@ -7,8 +7,10 @@
     "author": "webfactory GmbH <info@webfactory.de>",
     "license": "MIT",
     "devDependencies": {
-        "@actions/core": "^1.1.0",
-        "@zeit/ncc": "^0.20.5",
-        "child_process": "^1.0.2"
+        "@actions/core": "^1.2.4",
+        "@zeit/ncc": "^0.20.5"
+    },
+    "scripts": {
+        "build": "node scripts/build.js"
     }
 }

--- a/scripts/build.js
+++ b/scripts/build.js
@@ -1,0 +1,35 @@
+const { execSync } = require('child_process')
+const path = require('path')
+const fs = require('fs')
+
+const buildDir = path.join(process.cwd(), 'build')
+const distDir = path.join(process.cwd(), 'dist')
+
+const buildIndexJs = path.join(buildDir, 'index.js')
+const distIndexJs = path.join(distDir, 'index.js')
+const distCleanupJs = path.join(distDir, 'cleanup.js')
+
+if (!fs.existsSync(buildDir)) {
+    fs.mkdirSync(buildDir)
+}
+
+// Build the main index.js file
+console.log('Building index.js...')
+execSync(`./node_modules/.bin/ncc build index.js -q -o ${buildDir}`)
+if (fs.existsSync(distIndexJs)) {
+    fs.unlinkSync(distIndexJs)
+}
+fs.renameSync(buildIndexJs, distIndexJs)
+
+// Build the cleanup.js file
+console.log('Building cleanup.js...')
+execSync(`./node_modules/.bin/ncc build cleanup.js -q -o ${buildDir}`)
+if (fs.existsSync(distCleanupJs)) {
+    fs.unlinkSync(distCleanupJs)
+}
+fs.renameSync(buildIndexJs, distCleanupJs)
+
+console.log('Cleaning up...')
+fs.rmdirSync(buildDir)
+
+console.log('Done')

